File: shared/src/api_types/character_draft.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Atributos
// ---------------------------------------------------------------------------

/// Los seis atributos base de DnD 5.5e
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AttributesDto {
    pub strength: u32,
    pub dexterity: u32,
    pub constitution: u32,
    pub intelligence: u32,
    pub wisdom: u32,
    pub charisma: u32,
}

impl AttributesDto {
    /// Suma total — útil para validar point-buy (27 puntos estándar)
    pub fn point_buy_cost(&self) -> u32 {
        [
            self.strength,
            self.dexterity,
            self.constitution,
            self.intelligence,
            self.wisdom,
            self.charisma,
        ]
        .iter()
        .map(|&v| cost_for_score(v))
        .sum()
    }
}

/// Coste en puntos de point-buy según la tabla de 5.5e
fn cost_for_score(score: u32) -> u32 {
    match score {
        8 => 0,
        9 => 1,
        10 => 2,
        11 => 3,
        12 => 4,
        13 => 5,
        14 => 7,
        15 => 9,
        _ => u32::MAX, // valor inválido
    }
}

// ---------------------------------------------------------------------------
// Paso del wizard
// ---------------------------------------------------------------------------

/// En qué paso del wizard de creación se encuentra el draft.
/// El servidor lo actualiza en cada respuesta para que el cliente sepa qué mostrar.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum CreationStep {
    #[default]
    Name,
    Race,
    Class,
    Attributes,
    Background,
    Feats,
    Review,
    Complete,
}

impl CreationStep {
    pub fn next(&self) -> Self {
        match self {
            Self::Name => Self::Race,
            Self::Race => Self::Class,
            Self::Class => Self::Attributes,
            Self::Attributes => Self::Background,
            Self::Background => Self::Feats,
            Self::Feats => Self::Review,
            Self::Review => Self::Complete,
            Self::Complete => Self::Complete,
        }
    }

    pub fn is_complete(&self) -> bool {
        *self == Self::Complete
    }
}

// ---------------------------------------------------------------------------
// Draft del personaje — estado parcial durante la creación
// ---------------------------------------------------------------------------

/// Estado del personaje en construcción.
/// Se actualiza paso a paso desde el cliente y se persiste en el servidor.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CharacterDraft {
    /// ID del draft — asignado por el servidor al crear
    #[serde(skip_serializing_if = "Option::is_none")]
    pub draft_id: Option<Uuid>,

    /// Paso actual del wizard
    pub step: CreationStep,

    // --- Datos del personaje ---
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,

    /// ID de la raza elegida. Ej: "elf", "dragon_rider"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub race_id: Option<String>,

    /// ID de la clase elegida. Ej: "fighter", "wizard"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub class_id: Option<String>,

    /// ID del trasfondo elegido. Ej: "acolyte", "soldier"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub background_id: Option<String>,

    /// Atributos asignados (point-buy o método estándar)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributes: Option<AttributesDto>,

    /// IDs de los dones elegidos (puede ser vacío hasta el paso Feats)
    #[serde(default)]
    pub feat_ids: Vec<String>,

    /// Respuestas a los choices de raza/clase/trasfondo
    /// La clave es `"<entry_id>.<choice_id>"`. Ej: `"elf.subrace"` → `"high_elf"`
    #[serde(default)]
    pub choices: HashMap<String, serde_json::Value>,
}

// ---------------------------------------------------------------------------
// Requests del cliente
// ---------------------------------------------------------------------------

/// Inicia un nuevo draft. El servidor responde con el draft_id asignado.
#[derive(Debug, Deserialize)]
pub struct CreateDraftRequest {
    /// Nombre del jugador (no del personaje) — para asociar al WS
    pub player_name: String,
}

/// Avanza el wizard con los datos del paso actual.
/// El servidor valida, actualiza el draft y devuelve el siguiente paso.
#[derive(Debug, Deserialize)]
pub struct UpdateDraftRequest {
    pub draft_id: Uuid,
    pub step: CreationStep,

    /// Nombre del jugador — requerido solo en el último paso para el auto-save
    #[serde(skip_serializing_if = "Option::is_none")]
    pub player_name: Option<String>,

    // Campos opcionales — solo se envían en el paso correspondiente
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub race_id: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub class_id: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub background_id: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributes: Option<AttributesDto>,

    #[serde(default)]
    pub feat_ids: Vec<String>,

    #[serde(default)]
    pub choices: HashMap<String, serde_json::Value>,
}

// ---------------------------------------------------------------------------
// Respuestas del servidor
// ---------------------------------------------------------------------------

/// Respuesta a POST /character/draft/create y PUT /character/draft/:id
#[derive(Debug, Serialize)]
pub struct DraftResponse {
    pub draft: CharacterDraft,
    /// Errores de validación del paso actual, si los hay
    #[serde(default)]
    pub errors: Vec<String>,
    /// true si el draft fue finalizado y el personaje ya existe
    pub finalized: bool,
}

/// Respuesta a GET /character/draft/:id
#[derive(Debug, Serialize)]
pub struct DraftStatusResponse {
    pub draft: CharacterDraft,
    pub is_complete: bool,
}

File: dnd-movile/app/src/main/java/io/github/gasparkral/dnd/model/CharacterDraft.kt
package io.github.gasparkral.dnd.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

// ---------------------------------------------------------------------------
// Atributos
// ---------------------------------------------------------------------------

@Serializable
data class AttributesDto(
    val strength: Int,
    val dexterity: Int,
    val constitution: Int,
    val intelligence: Int,
    val wisdom: Int,
    val charisma: Int,
) {
    companion object {
        val DEFAULT = AttributesDto(8, 8, 8, 8, 8, 8)

        private val costTable = mapOf(
            8 to 0, 9 to 1, 10 to 2, 11 to 3,
            12 to 4, 13 to 5, 14 to 7, 15 to 9
        )

        fun costFor(score: Int): Int = costTable[score] ?: Int.MAX_VALUE
    }

    fun pointBuyCost(): Int = listOf(strength, dexterity, constitution, intelligence, wisdom, charisma)
        .sumOf { costFor(it) }

    fun withField(field: String, value: Int) = when (field) {
        "str" -> copy(strength = value)
        "dex" -> copy(dexterity = value)
        "con" -> copy(constitution = value)
        "int" -> copy(intelligence = value)
        "wis" -> copy(wisdom = value)
        "cha" -> copy(charisma = value)
        else  -> this
    }

    fun getField(field: String) = when (field) {
        "str" -> strength
        "dex" -> dexterity
        "con" -> constitution
        "int" -> intelligence
        "wis" -> wisdom
        "cha" -> charisma
        else  -> 8
    }
}

// ---------------------------------------------------------------------------
// Paso del wizard — espeja CreationStep de Rust
// ---------------------------------------------------------------------------

@Serializable
enum class CreationStep {
    @SerialName("name")       Name,
    @SerialName("race")       Race,
    @SerialName("class")      Class,
    @SerialName("attributes") Attributes,
    @SerialName("background") Background,
    @SerialName("feats")      Feats,
    @SerialName("review")     Review,
    @SerialName("complete")   Complete;

    fun isComplete() = this == Complete
}

// ---------------------------------------------------------------------------
// CharacterDraft — estado del personaje en construcción
// ---------------------------------------------------------------------------

@Serializable
data class CharacterDraft(
    @SerialName("draft_id")      val draftId: String? = null,
    val step: CreationStep = CreationStep.Name,
    val name: String? = null,
    @SerialName("race_id")       val raceId: String? = null,
    @SerialName("class_id")      val classId: String? = null,
    @SerialName("background_id") val backgroundId: String? = null,
    val attributes: AttributesDto? = null,
    @SerialName("feat_ids")      val featIds: List<String> = emptyList(),
    val choices: Map<String, JsonElement> = emptyMap(),
)

// ---------------------------------------------------------------------------
// Requests
// ---------------------------------------------------------------------------

@Serializable
data class CreateDraftRequest(
    @SerialName("player_name") val playerName: String,
)

@Serializable
data class UpdateDraftRequest(
    @SerialName("draft_id")      val draftId: String,
    val step: CreationStep,
    @SerialName("player_name")   val playerName: String? = null,
    val name: String? = null,
    @SerialName("race_id")       val raceId: String? = null,
    @SerialName("class_id")      val classId: String? = null,
    @SerialName("background_id") val backgroundId: String? = null,
    val attributes: AttributesDto? = null,
    @SerialName("feat_ids")      val featIds: List<String> = emptyList(),
    val choices: Map<String, JsonElement> = emptyMap(),
)

@Serializable
data class SetDraftStepRequest(
    val step: CreationStep,
)

// ---------------------------------------------------------------------------
// Responses
// ---------------------------------------------------------------------------

@Serializable
data class DraftResponse(
    val draft: CharacterDraft,
    val errors: List<String> = emptyList(),
    val finalized: Boolean = false,
)

@Serializable
data class DraftStatusResponse(
    val draft: CharacterDraft,
    @SerialName("is_complete") val isComplete: Boolean,
)
